src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.SymbolSet.cs

```diff
@@ -0,0 +1,268 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindSymbols.Finders;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.FindSymbols
+{
+    internal partial class FindReferencesSearchEngine
+    {
+        /// <summary>
+        /// Represents the set of symbols that the engine is searching for.  While the find-refs engine is passed an
+        /// initial symbol to find results for, the engine will often have to 'cascade' that symbol to many more symbols
+        /// that clients will also need.  This includes:
+        /// <list type="number">
+        /// <item>Cascading to all linked symbols for the requested symbol.  This ensures a unified set of results for a
+        /// particular symbol, regardless of what project context it was originally found in.</item>
+        /// <item>Symbol specific cascading.  For example, when searching for a named type, references to that named
+        /// type will be found through its constructors.</item>
+        /// <item>Cascading up and down the inheritance hierarchy for members (e.g. methods, properties, events). This
+        /// is controllable through the <see cref="FindReferencesSearchOptions.UnidirectionalHierarchyCascade"/>
+        /// option.</item>
+        /// </list>
+        /// </summary>
+        private abstract class SymbolSet
+        {
+            protected readonly FindReferencesSearchEngine Engine;
+
+            protected SymbolSet(FindReferencesSearchEngine engine)
+            {
+                Engine = engine;
+            }
+
+            protected Solution Solution => Engine._solution;
+
+            /// <summary>
+            /// Get a copy of all the symbols in the set.  Cannot be called concurrently with <see
+            /// cref="InheritanceCascadeAsync"/>
+            /// </summary>
+            public abstract ImmutableArray<ISymbol> GetAllSymbols();
+
+            /// <summary>
+            /// Update the set of symbols in this set with any appropriate symbols in the inheritance hierarchy brought
+            /// in within <paramref name="project"/>.  For example, given a project 'A' with interface <c>interface IGoo
+            /// { void Goo(); }</c>, and a project 'B' with class <c>class Goo : IGoo { public void Goo() { } }</c>,
+            /// then initially the symbol set will only contain IGoo.Goo.  However, when project 'B' is processed, this
+            /// will add Goo.Goo is added to the set as well so that references to it can be found.
+            /// </summary>
+            /// <remarks>
+            /// This method is non threadsafe as it mutates the symbol set instance.  As such, it should only be
+            /// called serially.  <see cref="GetAllSymbols"/> should not be called concurrently with this.
```

Comment:
would it be worth implementing these methods with a lock / some kind of assert in here verifying that no one is accidentally calling these methods concurrently, then delegate to other abstract methods?

Reply:
Yes. I can add a poor man's assert that only one is being called at a time.

Reply:
I like this idea. Maybe only in Debug though?